errors/errors.go

```diff
@@ -0,0 +1,195 @@
+	// categorized as any of the below cases.
+	InternalErr = NewRootError(0, "internal")
+
+	// UnauthorizedErr is used whenever a request without sufficient
+	// authorization is handled.
+	UnauthorizedErr = NewRootError(1, "unauthorized")
+
+	// NotFoundErr is used when a requested operation cannot be completed
+	// due to missing data.
+	NotFoundErr = NewRootError(2, "not found")
+
+	// InvalidMsgErr is returned whenever an event is invalid and cannot be
+	// handled.
+	InvalidMsgErr = NewRootError(3, "invalid message")
+
+	// InvalidModelErr is returned whenever a message is invalid and cannot
+	// be used (ie. persisted).
+	InvalidModelErr = NewRootError(4, "invalid model")
+)
+
+// NewRootError returns an error instance that should be used as the base for
+// creating error instances during runtime.
+//
+// Popular root errors are declared in this package, but extensions may want to
+// declare custom codes. This function ensures that no error code is used
+// twice. Attempt to reuse an error code results in panic.
+//
+// Use this function only during a program startup phase.
+func NewRootError(code uint32, description string) RootError {
+	if e, ok := usedCodes[code]; ok {
+		panic(fmt.Sprintf("error with code %d is already registered: %q", code, e.desc))
+	}
+	err := RootError{
+		code: code,
+		desc: description,
+	}
+	usedCodes[err.code] = err
+	return err
+}
+
+// usedCodes is keeping track of used codes to ensure uniqueness.
+var usedCodes = map[uint32]RootError{}
+
+// RootError represents a root error.
+//
+// Weave framework is using root error to categorize issues. Each instance
+// created during the runtime should wrap one of the declared root errors. This
+// allows error tests and returning all errors to the client in a safe manner.
+//
+// All popular root errors are declared in this package. If an extension has to
+// declare a custom root error, always use NewRootError function to ensure
+// error code uniqueness.
+type RootError struct {
+	code uint32
+	desc string
+}
+
+func (e RootError) Error() string    { return e.desc }
+func (e RootError) ABCICode() uint32 { return e.code }
+func (e RootError) ABCILog() string  { return e.desc }
+
+// New returns a new error. Returned instance is having the root cause set to
+// this error. Below two lines are equal
+//   e.New("my description")
+//   Wrap(e, "my description")
+func (e RootError) New(description string) error {
+	return Wrap(e, description)
+}
+
+// This Wrap implementation provides a transition layer between two Wrap
+// function implementations with incompatible notations.
+//
+// Once migration is complete, it will be removed and replaced by wrapng
+// function.
+func Wrap(err error, description ...string) TMError {
+	switch len(description) {
+	case 0:
+		// fmt.Fprintf
+		// debug.PrintStack()
+		return deprecatedLegacyWrap(err)
+	case 1:
+		return wrapng(err, description[0])
+	default:
+		panic("invalid Wrap notation used")
```

Comment:
Q: I do understand why you want to cover this case but it would be very confusing for me to find this as a root cause for a panic. Can we either just take the first description? Alternatively a Join would be ok for me too but this would give the wrong impression that multiple descr are allowed

Reply:
This was intended to be a transition interface. Although `func Wrap(err error, description ...string) TMError {` is very permissive, I want to ensure that only two of notations are supported (old and new one) and everything else must panic. 
This is not to be "used" and I think panic is ok. This is only because this is transition code, that must be removed when the rewrite is done. Keeping it strict should make removing it easier.

I see @ruseinov point and if we have time for this than I fully agree. This is a massive task though - on Friday I have spent a lot of time trying to convert just one extension and I have not finished yet. This task is not a simple find a replace.

Reply:
I do understand this api is for migration phase only but this panic comes without a need, IMHO.

Reply:
req: I would like this panic to be replaced by some default behaviour.  For me this is a blocker but I don't want to be dogmatic if majority is 👍 with it. Let's vote // cc @ruseinov

Reply:
I can change it to not panic. What should be the behavior instead?

Reply:
I'm happy with panic to be honest, I think the should fail fast and hard.

Reply:
But it is not fail fast, because Wrap is used at runtime. 
I had a similar case with gokit logger once. You don't think it is an issue unless it bites you hard on production. As I wrote in my first comment, either take the first description only or join them. It doesn't make a difference as we all assume this case should not happen. 🤞

Reply:
Why don’t we have two wraps then, one Wrap and one WrapWithDescription. Then after transition we can refactor all the usages via ide. I think the vararg approach is a concern cause it gives the wrong impresson and allows to produce code that would need to be migrated too.

Reply:
Why don’t we have two wraps then, one Wrap and one WrapWithDescription. Then after transition we can refactor all the usages via ide. I think the vararg approach is a concern cause it gives the wrong impresson and allows to produce code that would need to be migrated too.

Reply:
Why don’t we have two wraps then, one Wrap and one WrapWithDescription. Then after transition we can refactor all the usages via ide. I think the vararg approach is a concern cause it gives the wrong impresson and allows to produce code that would need to be migrated too.

Reply:
you can also use OldWrap and Wrap as it will serve exactly the same purpose. I don’t think combining everything into one method that is gonna change signature later is a good idea, refacotring names is automatic, signatures - not really